pkg/provisioning/changelog.go

```diff
@@ -0,0 +1,117 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provisioning
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+)
+
+// changelog should be adjusted every time we change the pipeline config and add
+// a new config version. Based on the changelog the parser will output warnings.
+var changelog = map[string][]change{
+	"1.0": {}, // no changes, initial version
+	"1.1": {{
+		field:      "pipelines.*.dead-letter-queue",
+		changeType: fieldIntroduced,
+		message:    "field dead-letter-queue was introduced in version 1.1, please update the pipeline config version",
+	}},
+}
+
+// change is a single change that was introduced in a specific version.
+type change struct {
+	// field contains the path to the field that was changed. Nested fields can
+	// be represented with dots (e.g. nested.field)
+	field      string
+	changeType changeType
+	// message is the log message that will be printed if a file is detected
+	// that uses this field with an unsupported version.
+	message string
+}
+
+// changeType defines the type of the change introduced in a specific version.
+type changeType int
+
+const (
+	fieldDeprecated changeType = iota
+	fieldIntroduced
```

Comment:
isn't deleted the same as deprecated? or do you mean deleted fields that are not supported anymore?

Reply:
Deleting a field would be a breaking change, ideally we shouldn't do that. We can deprecate a field and warn the user if they use it, but at the same time keep the behavior of that field if we encounter it. That's at least the current idea.

Reply:
Agreed that deleted configuration fields are not nice to work with. It would be nice if we write down our philosophy somewhere, even if it's a comment in the code.

Reply:
I'd argue that this is general advice we should follow in any publicly facing interface, from the API, the SDK interface, connector protocol, pipeline config files, CLI flags etc. For instance we did the same with the [required field](https://github.com/ConduitIO/conduit-connector-sdk/blob/e74f759f151935a9d6215eca3cef4fa7d503964f/specifier.go#L45-L47). It's just the way semantic versioning works, so I'm reluctant to add this comment here, and I'm not sure where the best place for this would be.